docs: Functions and triangle example (#849)
software-mansion/TypeGPU: apps/typegpu-docs/src/content/docs/fundamentals/functions.mdx

---
title: Functions
description: A guide on how to create and use the TypeGPU typed functions.
---

:::caution[Experimental]
Functions are an *unstable* feature. The API may be subject to change in the near future.
:::

:::note[Recommended reading]
We assume that you are familiar with the following concepts:
- <a href="https://webgpufundamentals.org/webgpu/lessons/webgpu-fundamentals.html" target="_blank" rel="noopener noreferrer">WebGPU Fundamentals</a>
- <a href="https://webgpufundamentals.org/webgpu/lessons/webgpu-wgsl.html" target="_blank" rel="noopener noreferrer">WebGPU Shading Language</a>
:::

TypeGPU allows writing shaders by composing typed functions, which are special wrappers around WGSL code. These functions can reference outside resources, like other user-defined or helper functions, buffers, bind group layouts etc. 

> TypeGPU also requires functions for quickly defining and running render and compute pipelines.

Reviewer: I don't think this line adds much.

## Creating a function

> The following code defines a function accepting one argument and returning one value.

Reviewer: ```suggestion
The following code defines a function that accepts one argument and returns one value.
```

```ts
const getColor = tgpu['~unstable']
  .fn([d.f32], d.vec4f)
  .does(/* wgsl */`(alpha: f32) -> vec4f {
    let color = vec4f(0.114, 0.447, 0.941, alpha);
    return color;
  }`)
  .$name('getColor');
```

> Functions are constructed by first defining their shells, that is their inputs and outputs. 

Reviewer: ```suggestion
Functions are constructed by first defining their shells, which specify their inputs and outputs.
```

Then the actual WGSL implementation is passed in through the `does` method.

If you're using Visual Studio Code, you can use an [extension](https://marketplace.visualstudio.com/items?itemName=ggsimm.wgsl-literal) that brings syntax highlighting to the code fragments marked with `/* wgsl */` comments.

## External resources

Functions can use external resources passed inside a record via the `$uses` method. 

> Externals can be any value or tgpu object that can be resolved to WGSL.

Reviewer: ```suggestion
Externals can be any value or TypeGPU resource that can be resolved to WGSL.
```
Also maybe we should be a little more specific here

```ts
const getBlue = = tgpu['~unstable']
  .fn({}, d.f32)
  .does(`() -> f32 { return 0.941 }`);

const getColor = tgpu['~unstable']
  .fn([d.f32], d.vec4f)
  .does(`(alpha: f32) -> vec4f {
    let color = vec4f(0.114, 0.447, getBlue(), alpha);
    return color;
  }`)
  .$uses({ getBlue })

>   .$name('getColor');

Reviewer: We could provide the resolution to showcase that the definitions are correctly added.

```

## Entry functions

> Defining entry functions is similar to regular ones, but is done through dedicated constructors: `tgpu['~unstable'].vertexFn`, `tgpu['~unstable'].fragmentFn`, `tgpu['~unstable'].computeFn`. 

Reviewer: ```suggestion
Defining entry functions is similar to regular ones, but is done through dedicated constructors:
- `tgpu['~unstable'].vertexFn`
- `tgpu['~unstable'].fragmentFn`
- `tgpu['~unstable'].computeFn`

```

They can be passed to root-defined pipelines and they accept special arguments like builtins (`d.builtin`) and decorated data (`d.location`).

```ts
const mainVertex = tgpu['~unstable']
  .vertexFn(
    { vertexIndex: d.builtin.vertexIndex },
    { outPos: d.builtin.position },
  )

>   .does(/* wgsl */ `(input: VertexInput) -> VertexOutput {

Reviewer: This behavior should be explained

    var pos = array<vec2f, 3>(
      vec2(0.0, 0.5),
      vec2(-0.5, -0.5),
      vec2(0.5, -0.5)
    );

    return VertexOutput(vec4f(pos[input.vertexIndex], 0.0, 1.0));
  }`);

const mainFragment = tgpu['~unstable']
  .fragmentFn({}, d.vec4f)
  .does(/* wgsl */ `() -> @location(0) vec4f {
    return getColor(1.0f);
  }`)
  .$uses({ getColor });
```

> After creating tgpu entry functions, you can access their input and output data schemas through dedicated properties:

Reviewer: I don't think that is consistent across all functions? Fragment only has output type and compute has neither.

```ts
mainVertex.inputType //  d.WgslStruct<WithLocations<{}>>

> mainVertex.outputType //  d.WgslStruct<WithLocations<{}>>

Reviewer: The type hints are not very helpful when all we have are builtins. Maybe add an additional argument? We could also explain why builtins are not here

---
title: Functions
description: A guide on how to create and use the TypeGPU typed functions.
---

:::caution[Experimental]
Functions are an *unstable* feature. The API may be subject to change in the near future.
:::

:::note[Recommended reading]
We assume that you are familiar with the following concepts:
- <a href="https://webgpufundamentals.org/webgpu/lessons/webgpu-fundamentals.html" target="_blank" rel="noopener noreferrer">WebGPU Fundamentals</a>
- <a href="https://webgpufundamentals.org/webgpu/lessons/webgpu-wgsl.html" target="_blank" rel="noopener noreferrer">WebGPU Shading Language</a>
:::

TypeGPU allows writing shaders by composing typed functions, which are special wrappers around WGSL code. 
These functions can reference outside resources, like other user-defined or helper functions, buffers, bind group layouts etc. 

## Creating a function

> The following code defines a function that accepts one argument and returns one value.
> 
> ```ts
> const getGradientColor = tgpu['~unstable']
>   .fn([d.f32], d.vec4f)
>   .does(`(ratio: f32) -> vec4f {
>     let color = mix(vec4f(0.769, 0.392, 1.0, 1), d.vec4f(0.114, 0.447, 0.941, 1), ratio);
>     return color;
>   }`)
>   .$name('getGradientColor');
> ```
> 
> Functions are constructed by first defining their shells, which specify their inputs and outputs.
> Then the actual WGSL implementation is passed in through the `does` method.

Reviewer: Very nitpicky but maybe reorder this? Flows better for me.
```suggestion
Functions are constructed by first defining their shells, which specify their inputs and outputs.
Then the actual WGSL implementation is passed in through the `does` method.

The following code defines a function that accepts one argument and returns one value.

```ts
const getGradientColor = tgpu['~unstable']
  .fn([d.f32], d.vec4f)
  .does(`(ratio: f32) -> vec4f {
    let color = mix(vec4f(0.769, 0.392, 1.0, 1), d.vec4f(0.114, 0.447, 0.941, 1), ratio);
    return color;
  }`)
  .$name('getGradientColor');
```
```

  .fragmentFn({ in: { uv: d.vec2f }, out: d.vec4f })
  .does(/* wgsl */ `(input: FragmentInput) -> @location(0) vec4f {
    return getGradientColor((input.uv[0] + input.uv[1]) / 2);
  }`)
  .$uses({ getGradientColor });
```

When entry function inputs or outputs are specified as objects containing builtins and inter-stage variables, the WGSL implementations need to access these arguments as passed in via structs (see *VertexInput*, *VertexOutput*, *FragmentInput*). 
TypeGPU schemas for these structs are created automatically by the library and their definitions are included when resolving the functions. The names referenced in the implementations do not matter.

## Usage in pipelines

Typed functions are crucial for simplified pipeline creation offered by TypeGPU. You can define and run pipelines as follows:

```ts
const pipeline = root['~unstable']
  .withVertex(mainVertex, {})
  .withFragment(mainFragment, { format: presentationFormat })
  .createPipeline();

pipeline
  .withColorAttachment({
    view: context.getCurrentTexture().createView(),
    clearValue: [0, 0, 0, 0],
    loadOp: 'clear',
    storeOp: 'store',
  })
  .draw(3);

root['~unstable'].flush();
```

> And the rendering result looks like this:

Reviewer: ```suggestion
The rendering result looks like this:
```

> ![rendering result - gradient triangle](../../../assets/triangle-result.png)

Reviewer 2: Could we put this asset next to the guide, and create a dedicated folder for both the guide and the asset?
```
content/docs/fundamentals/functions/index.mdx
content/docs/fundamentals/functions/triangle-result.png
```
